tracer/test/Datadog.Trace.Tests/CallTarget/IntegrationOptionsTests.cs

```diff
@@ -0,0 +1,41 @@
+﻿// <copyright file="IntegrationOptionsTests.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using Datadog.Trace.ClrProfiler.AutoInstrumentation.Http.HttpClient.HttpClientHandler;
+using Datadog.Trace.ClrProfiler.AutoInstrumentation.Testing.XUnit;
+using Datadog.Trace.ClrProfiler.CallTarget.Handlers;
+using Datadog.Trace.Configuration;
+using FluentAssertions;
+using Xunit;
+
+namespace Datadog.Trace.Tests.CallTarget
+{
+    public class IntegrationOptionsTests
+    {
+        [Fact]
+        public void CanGetIntegrationIdFromInstrumentAttribute()
+        {
+            var integrationType = typeof(HttpClientHandlerIntegration);
+            var targetType = typeof(System.Net.Http.HttpClientHandler);
+
+            var info = IntegrationOptions<object, object>.GetIntegrationId(integrationType, targetType);
+
+            info.Should().NotBeNull();
+            info.Value.Should().Be(IntegrationId.HttpMessageHandler);
+        }
+
+        [Fact]
+        public void CanGetIntegrationIdFromInstrumentAttributeWithMultipleAssemblyNames()
+        {
+            var integrationType = typeof(XUnitTestInvokerRunAsyncIntegration);
+            var targetType = typeof(Xunit.Sdk.TestInvoker<>);
+
+            var info = IntegrationOptions<object, object>.GetIntegrationId(integrationType, targetType);
+
+            info.Should().NotBeNull();
+            info.Value.Should().Be(IntegrationId.XUnit);
+        }
```

Comment:
I was curious how this would work with assembly-level attributes like `[assembly: CommandExecuteNonQuery(typeof(SystemDataSqlClientAdoNetClientData))]` in SqlClientDefinitions.cs. After adding the `System.Data.SqlClient` NuGet package for non-Framework builds, this is reflected in the following test code:

```csharp
        [Fact]
        public void CanGetIntegrationIdFromAssemblyLevelInstrumentAttribute()
        {
            var integrationType = typeof(CommandExecuteNonQueryIntegration);
            var targetType = typeof(System.Data.SqlClient.SqlCommand);

            var info = IntegrationOptions<object, object>.GetIntegrationId(integrationType, targetType);

            info.Should().NotBeNull();
            info.Value.Should().Be(IntegrationId.SqlClient);
        }
```

With this test code I found a small flaw, in that the attribute lookup would find the incorrect attribute (note the async difference):
`Datadog.Trace.ClrProfiler.AutoInstrumentation.AdoNet.AdoNetClientInstrumentMethodAttribute.CommandExecuteNonQueryAsyncAttribute`
versus
`Datadog.Trace.ClrProfiler.AutoInstrumentation.AdoNet.AdoNetClientInstrumentMethodAttribute.CommandExecuteNonQueryAttribute`

To find the matching attribute, we need to add one more conditional in `IntegrationOptions.GetIntegrationId`. I'll post that suggestion separately

Reply:
Note: If we add another NuGet package reference like `System.Data.SQLite.Core`, then another test case would be:
- Keep `integrationType` the same
- Assign `var targetType = typeof(System.Data.SQLite.SQLiteCommand);`
- Expect info to be `IntegrationId.Sqlite`

Reply:
One last unit test that we need to run is the following, which tests our abstract instrumentation:

```csharp
        [Fact]
        public void CanGetIntegrationIdFromAssemblyLevelInstrumentAttribute_IntegrationTypeIsDerived()
        {
            var integrationType = typeof(CommandExecuteNonQueryIntegration);
            var targetType = typeof(FakeCommand);

            var info = IntegrationOptions<object, object>.GetIntegrationId(integrationType, targetType);

            info.Should().NotBeNull();
            info.Value.Should().Be(IntegrationId.AdoNet);
        }
```

Our logic needs to handle the derived IntegrationType and it should always find the correct attribute. I can spend a little time trying to figure out what the logic is and share what I find

Reply:
I don't want to just commit all over your PR so here's what I found that works for IntegrationOptions.GetIntegrationId to go up some base types (includes the type's full name like I suggested earlier). What do you think of this approach?

```csharp
        // internal for testing
        internal static IntegrationId? GetIntegrationId(Type integrationType, Type targetType)
        {
            try
            {
                var attributes = integrationType.GetCustomAttributes(typeof(InstrumentMethodAttribute), inherit: true);
                if (attributes.Length > 0)
                {
                    var name = ((InstrumentMethodAttribute)attributes[0]).IntegrationName;
                    if (string.IsNullOrEmpty(name))
                    {
                        // shouldn't happen?
                        return null;
                    }

                    return IntegrationRegistry.TryGetIntegrationId(name, out var integrationId) ? integrationId : null;
                }

                // probably assembly attribute
                var assemblyAttributes = integrationType.Assembly.GetCustomAttributes(typeof(InstrumentMethodAttribute), inherit: true);
                if (assemblyAttributes.Length == 0)
                {
                    // can't work out which integration this is, shouldn't happen?
                    return null;
                }

                Type currentType = targetType;
                IntegrationId? result = GetIntegrationIdFromAssemblyAttributes(assemblyAttributes, integrationType.FullName, currentType, isDerived: false);
                if (result is not null)
                {
                    return result;
                }

                int maxDepth = 1;
                while (maxDepth > 0)
                {
                    currentType = currentType.BaseType;
                    result = GetIntegrationIdFromAssemblyAttributes(assemblyAttributes, integrationType.FullName, currentType, isDerived: true);
                    if (result is not null)
                    {
                        return result;
                    }

                    maxDepth--;
                }

                // not found
                return null;

                static IntegrationId? GetIntegrationIdFromAssemblyAttributes(object[] assemblyAttributes, string integrationTypeName, Type targetType, bool isDerived)
                {
                    var targetTypeName = targetType.FullName;
                    var targetAssemblyName = targetType.Assembly.GetName().Name;
                    var requiredIntegrationType = isDerived ? IntegrationType.Derived : IntegrationType.Default;

                    foreach (InstrumentMethodAttribute assemblyAttribute in assemblyAttributes)
                    {
                        if (assemblyAttribute.CallTargetIntegrationType == requiredIntegrationType && assemblyAttribute.TypeName == targetTypeName)
                        {
                            if (assemblyAttribute.AssemblyNames is not null)
                            {
                                foreach (var assemblyName in assemblyAttribute.AssemblyNames)
                                {
                                    if (assemblyName == targetAssemblyName && assemblyAttribute.CallTargetType.FullName == integrationTypeName)
                                    {
                                        return IntegrationRegistry.TryGetIntegrationId(assemblyAttribute.IntegrationName, out var integrationId) ? integrationId : null;
                                    }
                                }
                            }

                            if (assemblyAttribute.AssemblyName is not null
                             && assemblyAttribute.AssemblyName == targetAssemblyName
                             && assemblyAttribute.CallTargetType.FullName == integrationTypeName)
                            {
                                return IntegrationRegistry.TryGetIntegrationId(assemblyAttribute.IntegrationName, out var integrationId) ? integrationId : null;
                            }
                        }
                    }

                    return null;
                }
            }
            catch (Exception ex)
            {
                Log.Warning(ex, "Error determining associated integration name for CallTarget error");
                return null;
            }
        }
```

Reply:
I hope my comments aren't too nitpicky. I was spending a great deal of time thinking how the lookup could fail and I think with these considerations there are no more gaps

Reply:
Thanks for working through all this @zacharycmontoya! I actually wonder if there's a different approach we should be using here entirely, having to do reflection here feels really hacky to me.

I've just published #2288 which uses a source generator to generate the instrumentation definitions instead of a standalone nuke target. As part of that, I had to rework how we define the definitions entirely (we can no longer inherit from `[InstrumentMethod]`) so would have to rework this logic too.

But that also makes me wonder if we can source gen the "lookup" part here too, to avoid any of this ambiguity. I guess we should revisit that once we have a decision on #2288, before going too deep here

Reply:
> But that also makes me wonder if we can source gen the "lookup" part here too

That sounds ideal. Would you want to block this PR on that work or push this in as a good-enough first effort?

Reply:
Sounds good to me! I'll take a look at the source gen PR now

Reply:
I've rebased this on top of the source gen PR now, and updated the `GetIntegrationId()` code to be source generated instead in https://github.com/DataDog/dd-trace-dotnet/pull/2241/commits/e3bbf60344d00b7a5ea509e43ec20ef4b861565f. It's a lot simpler now, in that it's just a big switch statement. It makes a number of assumptions (e.g. two "normal" integrations of different `IntegrationId`s won't decorate the same class etc) but they all hold at the moment, and will cause compilation failures if they don't hold. I'll probably pull that into a separate PR once the source gen PR is approved and merged